### DISCUSSION TITLE
Hidden import not being processed

### DISCUSSION THREAD

**Discussion by machur** [pyinstaller/pyinstaller#7762] [PyInstaller] [answered]
Hi, I'm trying to pack an application for visualization using pyinstaller. Under the hood VTK is used for rendering. There some missing imports that I was able to provide via hidden imports, but one of the imports is not being processed, even though is in the cmd/spec file:

```
a = Analysis(
    ['main_window.py'],
    pathex=[],
    binaries=[],
    datas=[],
    hiddenimports=['vtkmodules.vtkIOXML', 'vtkmodules.vtkCommonMisc', 'vtkmodules.vtkIOCore'],
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)
```

I'm getting the following error:

> ImportError: Failed to load vtkIOInfovis: No module named vtkmodules.vtkIOXML
> [9156] Failed to execute script 'main_window' due to unhandled exception!

Pyinstaller suppose to look for three hidden imports (vtkmodules.vtkIOXML, vtkmodules.vtkCommonMisc, vtkmodules.vtkIOCore), but only two have been analysed according to the logs:

> ...
> 27069 INFO: Loading module hook 'hook-sqlite3.py' from 'C:\\Users\\machur\\.conda\\envs\\heart\\lib\\site-packages\\PyInstaller\\hooks'...
> 28179 INFO: Analyzing hidden import 'vtkmodules.vtkCommonMisc'
> 28179 INFO: Analyzing hidden import 'vtkmodules.vtkIOCore'
> 28179 INFO: Processing module hooks...
> 28851 INFO: Loading module hook 'hook-setuptools.py' from 'C:\\Users\\machur\\.conda\\envs\\heart\\lib\\site-packages\\PyInstaller\\hooks'...
> ...

Would you please advise?

**Comment by bwoodsend** [comment, marked as answer]
It's not being analysed as a hidden import because it wasn't hidden - PyInstaller had already found and collected it. The issue here is that these `ModuleNotFoundError`s are blaming `vtkmodules.vtkIOXML` when it's actually its dependencies that are missing. If you try to import `vtkmodules.vtkIOXML` directly, the error tells you the real subpackage that is missing.

It works for me with these hidden imports.

```python
['vtkmodules.vtkCommonCore', 'vtkmodules.vtkCommonMath', 'vtkmodules.vtkCommonTransforms', 'vtkmodules.vtkCommonDataModel', 'vtkmodules.vtkCommonExecutionModel', 'vtkmodules.vtkIOCore', 'vtkmodules.vtkIOLegacy', 'vtkmodules.vtkIOXMLParser', 'vtkmodules.vtkIOXML', 'vtkmodules.vtkIOInfovis']
```

This is a lot worse than it used to be. I remember never needing more than a handful of easy to figure out hidden imports for VTK.

**Reply by machur** [reply]
@bwoodsend thanks for the quick response, that's super helpful! I had to add a couple of more imports, but your method worked like a charm.

BTW I work either in ML and struggle with CUDA, or I work in visualization and struggle with VTK, so I don't expect anything to go smoothly, there's always a long list of issues with these two.
